compiler/rustc_borrowck/src/session_diagnostics.rs

```diff
@@ -240,3 +304,546 @@ pub(crate) struct MoveBorrow<'a> {
+        #[subdiagnostic(eager)]
+        eager_label: MutMultiLoopLabel<'a>,
+    },
+    #[diag(borrowck_cannot_mutably_borrow_multiply, code = "E0499")]
+    ChangedSpan {
+        new_place_name: &'a str,
+        place: &'a str,
+        old_place: &'a str,
+        is_place_empty: bool,
+        is_old_place_empty: bool,
+        #[primary_span]
+        #[label(second_mut_borrow_label)]
+        new_loan_span: Span,
+        #[label]
+        old_loan_span: Span,
+        #[label(first_mut_end_label)]
+        old_load_end_span: Option<Span>,
+    },
+}
+
+#[derive(Subdiagnostic)]
+#[label(borrowck_mutably_borrow_multiply_loop_label)]
+pub(crate) struct MutMultiLoopLabel<'a> {
+    pub new_place_name: &'a str,
+    pub place: &'a str,
+    pub is_place_empty: bool,
+    #[primary_span]
+    pub new_loan_span: Span,
+}
+
+#[derive(Diagnostic)]
+#[diag(borrowck_cannot_uniquely_borrow_by_one_closure, code = "E0500")]
+pub(crate) struct ClosureUniquelyBorrowErr<'a> {
+    #[primary_span]
+    #[label]
+    pub new_loan_span: Span,
+    pub container_name: &'a str,
+    pub desc_new: &'a str,
+    pub opt_via: &'a str,
+    #[label(occurs_label)]
+    pub old_loan_span: Span,
+    pub noun_old: &'a str,
+    pub old_opt_via: &'a str,
+    #[label(ends_label)]
+    pub previous_end_span: Option<Span>,
+}
+
+#[derive(Diagnostic)]
+#[diag(borrowck_cannot_reborrow_already_uniquely_borrowed, code = "E0501")]
+pub(crate) struct ClosureReBorrowErr<'a> {
+    #[primary_span]
+    #[label]
+    pub new_loan_span: Span,
+    pub container_name: &'a str,
+    pub desc_new: &'a str,
+    pub opt_via: &'a str,
+    pub kind_new: &'a str,
+    #[label(occurs_label)]
+    pub old_loan_span: Span,
+    pub old_opt_via: &'a str,
+    #[label(ends_label)]
+    pub previous_end_span: Option<Span>,
+    pub second_borrow_desc: &'a str,
+}
+
+#[derive(Subdiagnostic)]
+pub(crate) enum BorrowOccurLabel<'a> {
+    #[label(borrowck_borrow_occurs_here)]
+    Here {
+        #[primary_span]
+        span: Span,
+        kind: &'a str,
+    },
+    #[label(borrowck_borrow_occurs_here_overlap)]
+    HereOverlap {
+        #[primary_span]
+        span: Span,
+        kind_new: &'a str,
+        msg_new: &'a str,
+        msg_old: &'a str,
+    },
+    #[label(borrowck_borrow_occurs_here_via)]
+    HereVia {
+        #[primary_span]
+        span: Span,
+        kind_old: &'a str,
+        is_msg_old_empty: bool,
+        msg_old: &'a str,
+    },
+}
+
+#[derive(Diagnostic)]
+#[diag(borrowck_cannot_reborrow_already_borrowed, code = "E0502")]
+pub(crate) struct ReborrowBorrowedErr<'a> {
+    pub desc_new: &'a str,
+    pub is_msg_new_empty: bool,
+    pub msg_new: &'a str,
+    pub kind_new: &'a str,
+    pub noun_old: &'a str,
+    pub kind_old: &'a str,
+    pub is_msg_old_empty: bool,
+    pub msg_old: &'a str,
+    #[primary_span]
+    pub span: Span,
+    #[label]
+    pub old_load_end_span: Option<Span>,
+    #[subdiagnostic(eager)]
```

Comment:
broke by #105233 @mejrs

Reply:
I agree something like `BorrowOccurLabel` would benefit from `(eager)`. let me see if I can put that back...

As an aside, looking at the `cannot_reborrow_already_borrowed` function, I see that it has only three call locations, and they either invoke this with (immutable, mutable) or (mutable, immutable). You can simplify `ReborrowBorrowedErr` a lot if you split it up into `CannotBorrowMutAlreadyBorrowedImmutErr` and `CannotBorrowImmutAlreadyBorrowedMutErr`.

Reply:
my flawful brain failed to parse `CannotBorrowMutAlreadyBorrowedImmutErr`..

anyway, I could put `(eager)` back if you don't feel like it, been messing around with macro a lot lately

Reply:
I've opened a PR :)